### PR TITLE
Feat/sma 540 check paymaster userop

### PR DESCRIPTION
**Pull request by VGabriel45**
# Summary

Refactor logic for ERC20 paymaster user op validation in setPaymasterUserOp function + tests

## Change Type

- [ ] Bug Fix
- [x] Refactor
- [ ] New Feature
- [ ] Breaking Change
- [ ] Documentation Update
- [ ] Performance Improvement
- [ ] Other

# Checklist

- [x] My code follows this project's style guidelines
- [x] I've reviewed my own code
- [ ] I've added comments for any hard-to-understand areas
- [ ] I've updated the documentation if necessary
- [x] My changes generate no new warnings
- [x] I've added tests that prove my fix is effective or my feature works
- [x] All unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published

### CODE REVIEW

**Review by joepegler — COMMENTED** [review]
Can we make sure that everything in here can be exported from `@biconomy/account`, and update the import here? 

https://github.com/bcnmy/biconomy-client-sdk/pull/384/files#diff-f9b56b02ea191c9a2bf7380a0dd2702bd35827ffefc83014b18bf2d03a904f76L6

File: packages/account/tests/account.e2e.spec.ts [hunk 1]
```diff
@@ -1,9 +1,15 @@
 import { TestData } from "../../../tests";
-import { PaymasterUserOperationDto, createSmartAccountClient } from "../src/index";
+import {
+  PaymasterUserOperationDto,
+  createSmartAccountClient,
+  FeeQuotesOrDataResponse,
+  IHybridPaymaster,
+  PaymasterFeeQuote,
+  PaymasterMode,
+} from "../src/index";
```
Comment (livingrockrises) [inline packages/account/tests/account.e2e.spec.ts:9]: we moved form paymaster package to this package?

File: packages/account/tests/account.e2e.spec.ts [hunk 2]
```diff
@@ -166,6 +178,50 @@ describe("Account Tests", () => {
     expect(newBalance - balance).toBe(1n);
   }, 60000);
 
+  it("Should throw and error if missing field for ERC20 Paymaster user op", async () => {
```
Comment (livingrockrises) [inline packages/account/tests/account.e2e.spec.ts:181]: does this mean buildUserOp would also throw?

**Comment by joepegler** [comment]
setPaymasterUserOp sounds like it’s setting something in place. getPaymasterUserOp seems a more appropriate name. wdyt?

File: packages/account/tests/account.e2e.spec.ts (hunk 1 above)
Comment (joepegler) [reply packages/account/tests/account.e2e.spec.ts:9]: We did this when we abstracted away the paymaster -> when we build it from the biconomyPaymasterApiKey

File: packages/account/src/BiconomySmartAccountV2.ts [hunk 3]
```diff
@@ -498,11 +498,58 @@ export class BiconomySmartAccountV2 extends BaseSmartContractAccount {
   }
 
   /**
-   * Sets paymaster-related fields in the provided user operation based on the specified paymaster service data.
    *
-   * @param {Partial<UserOperationStruct>} userOp - The partial user operation structure to be modified.
-   * @param {PaymasterUserOperationDto} paymasterServiceData - The paymaster service data containing mode and additional information.
-   * @returns {Promise<Partial<UserOperationStruct>>} A promise that resolves to the modified user operation structure.
+   * @param userOp
```
Comment (joepegler) [inline packages/account/src/BiconomySmartAccountV2.ts:502]: Thanks 👍

File: packages/account/tests/account.e2e.spec.ts (hunk 2 above)
Comment (VGabriel45) [reply packages/account/tests/account.e2e.spec.ts:181]: Yes, if ERC20 mode is passed without all required fields

File: packages/account/tests/account.e2e.spec.ts (hunk 1 above)
Comment (livingrockrises) [reply packages/account/tests/account.e2e.spec.ts:9]: I see. confused because I thought it was already merged in v4

**Comment by livingrockrises** [comment]
> > setPaymasterUserOp sounds like it’s setting something in place. getPaymasterUserOp seems a more appropriate name. wdyt?
> 
> True, fixed

or patchPaymasterAndData
wdyt?

**Comment by livingrockrises** [comment]
> > > setPaymasterUserOp sounds like it’s setting something in place. getPaymasterUserOp seems a more appropriate name. wdyt?
> > 
> > 
> > True, fixed
> 
> or patchPaymasterAndData wdyt?

or patchUserOperation

File: packages/account/tests/account.e2e.spec.ts (hunk 1 above)
Comment (joepegler) [reply packages/account/tests/account.e2e.spec.ts:9]: These are just some additional types / interfaces coming from the same Paymaster package. If we've made the decision to import that Paymaster class already to help with defaulting, then re-exporting both it and its relevant types and interfaces from accounts makes sense, so there's a central package to import stuff from. 

v4 has moved us in the direction of having only two user-facing packages: accounts + modules. Everything else will  just be imported into those packages.

File: packages/account/tests/account.e2e.spec.ts [hunk 4]
```diff
@@ -166,6 +178,50 @@ describe("Account Tests", () => {
...
+    const transaction = {
+      to: nftAddress, // NFT address
+      data: encodedCall,
+    };
+
+    const userOp = await smartWallet.buildUserOp([transaction]);
+
+    const feeQuotesResponse: FeeQuotesOrDataResponse = await (
+      smartWallet.paymaster as IHybridPaymaster<PaymasterUserOperationDto>
+    ).getPaymasterFeeQuotesOrData(userOp, {
```
Comment (AmanRaj1608) [inline packages/account/tests/account.e2e.spec.ts:210]: wish we could solve this also by adding another tokenTaymaster instance or something in SDK

**Review by AmanRaj1608 — APPROVED** [review]

**Comment by AmanRaj1608** [comment]
`getPaymasterUserOp` or `patchUserOperation` both sounds good.

File: packages/account/tests/account.e2e.spec.ts (hunk 4 above)
Comment (livingrockrises) [reply packages/account/tests/account.e2e.spec.ts:210]: yes I mentioned [this](https://github.com/bcnmy/biconomy-client-sdk/issues/303) in dsm yesterday

**Review by livingrockrises — APPROVED** [review]

**Pull request merged**